src/main/kotlin/com/ampnet/web3provider/repository/RedisRepositoryImpl.kt

```diff
@@ -1,24 +1,41 @@
 package com.ampnet.web3provider.repository
 
+import com.ampnet.web3provider.controller.pojo.JsonRpcRequest
+import com.ampnet.web3provider.controller.pojo.ProviderResponse
 import com.ampnet.web3provider.enums.RedisEntity
+import com.ampnet.web3provider.service.DefaultProviderService
 import org.springframework.data.redis.core.HashOperations
 import org.springframework.data.redis.core.RedisTemplate
 import org.springframework.stereotype.Repository
 import java.time.Duration
 
 @Repository
-class RedisRepositoryImpl(private val redisTemplate: RedisTemplate<String, String>) : RedisRepository {
+class RedisRepositoryImpl(
+    private val redisTemplate: RedisTemplate<String, Any>,
+    private val defaultProviderService: DefaultProviderService
+) : RedisRepository {
 
-    private val hashOperations: HashOperations<String, String, String> = redisTemplate.opsForHash()
+    private val hashOperations: HashOperations<String, String, Any> = redisTemplate.opsForHash()
 
-    override fun getCache(key: String, hashKey: String): String? = hashOperations.get(key, hashKey)
+    override fun getCache(key: String, hashKey: String): Any? = hashOperations.get(key, hashKey)
 
-    override fun updateCache(redisEntity: RedisEntity, hashKey: String, value: String) {
+    override fun updateCache(redisEntity: RedisEntity, hashKey: String, value: Any) {
         hashOperations.put(redisEntity.methodName, hashKey, value)
         redisTemplate.expire(redisEntity.methodName, Duration.ofSeconds(redisEntity.ttlInSec))
     }
 
     override fun deleteCache(key: String, hashKey: String) {
         hashOperations.delete(key, hashKey)
     }
+
+    override fun getResponseFromCacheOrProvider(
+        entity: RedisEntity,
+        hashKey: String,
+        request: JsonRpcRequest
+    ): ProviderResponse {
+        getCache(entity.methodName, hashKey)?.let { return ProviderResponse(request, it) }
+        val providerResponse = defaultProviderService.getResponse(request)
+        providerResponse.result?.let { updateCache(entity, hashKey, it) }
+        return providerResponse
+    }
```

Comment:
I wouldn't couple cache with provider service. The cache should just return value if it is stored in cache or return null. Other service should know how what to do on empty value.

Reply:
If not available from cache get it from default provider and update the cache. It's repeated in each service. I can extract it in a ServiceUtil method?
```
internal object ServiceUtils {

    fun getResponseFromCacheOrProvider(
        defaultProviderService: DefaultProviderService,
        redisRepository: RedisRepository,
        data: ProviderData
    ): ProviderResponse {
        redisRepository.getCache(data.entity.methodName, data.hashKey)?.let {
            return ProviderResponse(data.request, it)
        }
        val providerResponse = defaultProviderService.getResponse(data.request)
        providerResponse.result?.let { redisRepository.updateCache(data.entity, data.hashKey, it) }
        return providerResponse
    }
}

data class ProviderData (
    val entity: RedisEntity,
    val hashKey: String,
    val request: JsonRpcRequest
)
```

Reply:
Why not move caching `provider response` inside `provider service`?

Reply:
Then it couples provider service with caching. Currently my controller checks if I have implemented a method, otherwise it just fetches the value from the provider. Ok, I could make another method in my provider to fetch and update cache as well.

Reply:
Of course, you have to couple things somewhere, but it makes more sense to couple provider service to cache than the other way around. The cache is simpler, lower in the hierarchy and this way it is simpler to plug and change if needed.